crates/ra_editor/src/line_index_utils.rs

```diff
@@ -0,0 +1,396 @@
+        self.line += 1;
+        self.last_newline = newline;
+        self.col_adjust = TextUnit::from(0);
+    }
+
+    fn adjust_col(&mut self, range: &TextRange) {
+        self.col_adjust += range.len() - TextUnit::from(1);
+    }
+}
+
+pub fn translate_offset_with_edit(
+    line_index: &LineIndex,
+    offset: TextUnit,
+    edits: &[AtomTextEdit],
+) -> LineCol {
+    let mut sorted_edits: Vec<&AtomTextEdit> = Vec::with_capacity(edits.len());
+    for edit in edits {
+        let insert_index =
+            sorted_edits.upper_bound_by_key(&edit.delete.start(), |x| x.delete.start());
+        sorted_edits.insert(insert_index, &edit);
+    }
+
+    let mut state = Edits::new(&sorted_edits);
+
+    let mut res = RunningLineCol::new();
+
+    macro_rules! test_step {
+        ($x:ident) => {
+            match &$x {
+                Step::Newline(n) => {
+                    if offset < *n {
+                        return res.to_line_col(offset);
+                    } else {
+                        res.add_line(*n);
+                    }
+                }
+                Step::Utf16Char(x) => {
+                    if offset < x.end() {
+                        // if the offset is inside a multibyte char it's invalid
+                        // clamp it to the start of the char
+                        let clamp = offset.min(x.start());
+                        return res.to_line_col(clamp);
+                    } else {
+                        res.adjust_col(x);
+                    }
+                }
+            }
+        };
+    }
+
+    for orig_step in LineIndexStepIter::from(line_index) {
+        loop {
+            let translated_step = state.translate_step(&orig_step);
+            match state.next_step(&translated_step) {
+                NextNewlines::Use => {
+                    test_step!(translated_step);
+                    break;
+                }
+                NextNewlines::ReplaceMany(ns) => {
+                    for n in ns {
+                        test_step!(n);
+                    }
+                    break;
+                }
+                NextNewlines::AddMany(ns) => {
+                    for n in ns {
+                        test_step!(n);
+                    }
+                }
+            }
+        }
+    }
+
+    loop {
+        match state.next_inserted_newlines() {
+            None => break,
+            Some(ns) => {
+                for n in ns {
+                    test_step!(n);
+                }
+            }
+        }
+    }
+
+    res.to_line_col(offset)
+}
+
+/// Simplest implementation to use as reference in proptest and benchmarks
+pub fn translate_after_edit(
+    pre_edit_text: &str,
+    offset: TextUnit,
+    edits: Vec<AtomTextEdit>,
+) -> LineCol {
+    let text = edit_text(pre_edit_text, edits);
+    line_index::to_line_col(&text, offset)
+}
+
+fn edit_text(pre_edit_text: &str, mut edits: Vec<AtomTextEdit>) -> String {
```

Comment:
This could be `TextEdit::apply` I think?

Reply:
Yes, I will change it